### PR TITLE
Separate most of rustc::lint::builtin into a separate crate.

### PR DESCRIPTION
**Pull request by huonw**
This pulls out the implementations of most built-in lints into a
separate crate, to reduce edit-compile-test iteration times with
librustc_lint and increase parallelism. This should enable lints to be
refactored, added and deleted much more easily as it slashes the
edit-compile cycle to get a minimal working compiler to test with (`make
rustc-stage1`) from

```
librustc -> librustc_typeck -> ... -> librustc_driver ->
    libcore -> ... -> libstd
```

to

```
librustc_lint -> librustc_driver -> libcore -> ... libstd
```

which is significantly faster, mainly due to avoiding the librustc build
itself.

The intention would be to move as much as possible of the infrastructure
into the crate too, but the plumbing is deeply intertwined with librustc
itself at the moment. Also, there are lints for which diagnostics are
registered directly in the compiler code, not in their own crate
traversal, and their definitions have to remain in librustc.

This is a [breaking-change] for direct users of the compiler APIs:
callers of `rustc::session::build_session` or
`rustc::session::build_session_` need to manually call
`rustc_lint::register_builtins` on their return value.

This should make #22206 easier.

### CODE REVIEW

**Comment by nagisa** [comment]
Nice improvement!

Since we have a separate crate for lints now, would be nice to split librustc_lint/builtin.rs into several files, e.g. each for a lint.

**Comment by huonw** [comment]
@nagisa yeah definitely, that's #22206. I don't have time to do it all immediately, but this patch makes it much easier to do incrementally since the compile times are much shorter. I thought it better to land this (to let everyone get the benefits now) than have it blocked on me doing the reorganisation first.

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -0,0 +1,2103 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Lints in the Rust compiler.
+//!
+//! This contains lints which can feasibly be implemented as their own
+//! AST visitor. Also `rustc::lint::builtin`, which contains the
```
Comment (kmcallister) [inline src/librustc_lint/builtin.rs]: Should be "Also see"?

File: src/librustc_lint/lib.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    macro_rules! add_lint_group {
+        ($sess:ident, $name:expr, $($lint:ident),*) => (
+            store.register_group($sess, false, $name, vec![$(LintId::of(builtin::$lint)),*]);
+            )
+    }
+
+    add_builtin!(sess,
+                 HardwiredLints,
```
Comment (kmcallister) [inline src/librustc_lint/lib.rs]: `HardwiredLints` stayed in librustc, right?

File: src/librustc_lint/lib.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                    UNUSED_MUT, UNREACHABLE_CODE, UNUSED_MUST_USE,
+                    UNUSED_UNSAFE, PATH_STATEMENTS);
+
+    // We have one lint pass defined specially
+    store.register_pass(sess, false, box lint::GatherNodeLevels as LintPassObject);
+
+    // Insert temporary renamings for a one-time deprecation
+    store.register_renamed("raw_pointer_deriving", "raw_pointer_derive");
+
+    store.register_renamed("unknown_features", "unused_features");
```
Comment (kmcallister) [inline src/librustc_lint/lib.rs]: Can these go away now?

File: src/librustc_lint/lib.rs (hunk 2 above)
Comment (huonw) [inline src/librustc_lint/lib.rs]: Yes

File: src/librustc_lint/lib.rs (hunk 3 above)
Comment (huonw) [inline src/librustc_lint/lib.rs]: I'm happy to remove them, but they're very low maintenance and it is helpful for people updating. *shrug*

File: src/librustc_lint/lib.rs (hunk 2 above)
Comment (kmcallister) [inline src/librustc_lint/lib.rs]: Does that mean it shouldn't be listed here?

Comment (huonw) [inline src/librustc_lint/lib.rs]: This function does all the registration, librustc just has the declaration of the type/constants.

File: src/librustc_lint/builtin.rs (hunk 1 above)
Comment (huonw) [inline src/librustc_lint/builtin.rs]: Yup

**Comment by huonw** [comment]
Rebased & fixed the missing word.

**Comment by Manishearth** [comment]
\o/

**Pull request merged**